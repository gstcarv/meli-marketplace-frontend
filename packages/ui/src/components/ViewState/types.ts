import { PropsWithChildren } from 'react';

export type ViewStateProps = PropsWithChildren<{
    /**
     * If is error
     */
    isError?: boolean;

    /**
     * If is empty
     */
    isEmpty?: boolean;

    /**
     * If is empty
     */
    isLoading?: boolean;
}>;
