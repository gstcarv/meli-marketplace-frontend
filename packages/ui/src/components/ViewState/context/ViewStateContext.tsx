import { createContext } from 'react';
import { ViewStateProps } from '../types';

export type ViewStateContextType = Pick<ViewStateProps, 'isEmpty' | 'isError' | 'isLoading'>;

export const ViewStateContext = createContext<ViewStateContextType>({} as ViewStateContextType);
