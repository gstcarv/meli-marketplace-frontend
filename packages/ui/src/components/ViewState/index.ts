export { ViewState } from './ViewState';
export { type ViewStateProps } from './types';
export { type ViewStateStatusProps } from './components/ViewStateStatus/types';
