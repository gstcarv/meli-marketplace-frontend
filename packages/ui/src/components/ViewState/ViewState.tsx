import { createViewStateStatus } from './components/ViewStateStatus';
import { ViewStateContext } from './context/ViewStateContext';
import { ViewStateProps } from './types';

export const ViewState = ({ children, ...props }: ViewStateProps) => (
    <ViewStateContext.Provider value={props}>{children}</ViewStateContext.Provider>
);

ViewState.Loading = createViewStateStatus((ctx) => ctx.isLoading);

ViewState.Error = createViewStateStatus((ctx) => ctx.isError);

ViewState.Empty = createViewStateStatus((ctx) => ctx.isEmpty);

ViewState.Success = createViewStateStatus((ctx) => !ctx.isEmpty && !ctx.isError && !ctx.isLoading);
