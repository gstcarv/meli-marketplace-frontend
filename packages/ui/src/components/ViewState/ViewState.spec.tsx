import { render, screen } from '@testing-library/react';
import { ViewStateContextType } from './context/ViewStateContext';
import { ViewState } from './ViewState';

describe('<ViewState />', () => {
    const ComponentStub = (props: ViewStateContextType) => (
        <ViewState {...props}>
            <ViewState.Empty>empty-content</ViewState.Empty>

            <ViewState.Error>error-content</ViewState.Error>

            <ViewState.Loading>loading-content</ViewState.Loading>

            <ViewState.Success>success-content</ViewState.Success>
        </ViewState>
    );

    it('should render loading', () => {
        const { container } = render(<ComponentStub isLoading />);

        expect(screen.getByText('loading-content')).toBeInTheDocument();
        expect(screen.queryByText('success-content')).not.toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    it('should render error', () => {
        const { container } = render(<ComponentStub isError />);

        expect(screen.getByText('error-content')).toBeInTheDocument();
        expect(screen.queryByText('success-content')).not.toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    it('should render empty', () => {
        const { container } = render(<ComponentStub isEmpty />);

        expect(screen.getByText('empty-content')).toBeInTheDocument();
        expect(screen.queryByText('success-content')).not.toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    it('should render success', () => {
        const { container } = render(<ComponentStub />);

        expect(screen.queryByText('success-content')).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });
});
