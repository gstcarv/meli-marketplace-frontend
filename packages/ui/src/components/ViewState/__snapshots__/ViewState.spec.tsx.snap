// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ViewState /> should render empty 1`] = `
<div>
  empty-content
</div>
`;

exports[`<ViewState /> should render error 1`] = `
<div>
  error-content
</div>
`;

exports[`<ViewState /> should render loading 1`] = `
<div>
  loading-content
</div>
`;

exports[`<ViewState /> should render success 1`] = `
<div>
  success-content
</div>
`;
